CFI improvements to the AArch64 fiber implementation

akirilov-arm opened the pull request:
Now the fiber implementation on AArch64 authenticates function
return addresses and includes the relevant BTI instructions, except
on macOS.

Copyright (c) 2022, Arm Limited.

<!--

Please ensure that the following steps are all taken care of before submitting
the PR.

- [ ] This has been discussed in issue #..., or if not, please tell us why
  here.
- [ ] A short description of what this does, why it is needed; if the
  description becomes long, the matter should probably be discussed in an issue
  first.
- [ ] This PR contains test cases, if meaningful.
- [ ] A reviewer from the core maintainer team has been assigned for this PR.
  If you don't know who could review this, please indicate so. The list of
  suggested reviewers on the right can help you.

Please ensure all communication adheres to the [code of
conduct](https://github.com/bytecodealliance/wasmtime/blob/master/CODE_OF_CONDUCT.md).
-->

alexcrichton commented on crates/cranelift/src/lib.rs:
```diff
@@ -217,7 +217,22 @@ fn func_signature(
+            if on_apple_aarch64 {
+                // FIXME: We need an Apple-specific calling convention, so that
+                // Cranelift's ABI implementation generates unwinding directives
+                // about pointer authentication usage.
```
Is the "FIXME" here referring to that theoretically we should add like a `FastAppleAarch64` convention or something like that? If that's the case it's definitely ok to leave this just as a comment for now, and perosnally I'd be ok leaving this as an implementation detial until someone otherwise proves that it's better to add the fast aarch64 variant.

akirilov-arm replied:
Yes, that's right. TBH currently the ABI implementation does not really distinguish between calling conventions, i.e. it generates essentially the same code in all cases (except for the unwinding information, as discussed here).

alexcrichton commented on crates/fiber/src/arch/aarch64.S:
```diff
@@ -10,12 +10,29 @@
+    .word 3; // GNU_PROPERTY_AARCH64_FEATURE_1_BTI | GNU_PROPERTY_AARCH64_FEATURE_1_PAC
+    .word 0;
+    .popsection
+#endif
```
Personally I'm pretty unfamiliar with all of this. Would you be up for adding a comment here to why this is necessary and perhaps an online doc if possible explaining the format of this section?

alexcrichton commented on crates/fiber/src/arch/aarch64.S:
```diff
@@ -10,12 +10,29 @@
 TYPE(wasmtime_fiber_switch)
 FUNCTION(wasmtime_fiber_switch):
+#ifndef CFG_TARGET_OS_macos
+    hint #25 // paciasp
```
To confirm, this is because the macos assembler doesn't have support for the `paciasp` instruction natively? (do you know if there's a version that does support it?)

alexcrichton replied:
Oh wow sorry I completely misread "ifndef CFG_TARGET_OS_macos" as "if macos" which is entirely the opposite of what's happening here. 

Otherwise those makes sense! Although the same question still somewhat applies, do general assemblers not recognize `paciasp` as an instruction? (or is there like some common assembler that only recently got support for such syntax?)

cfallin replied:
At least for `bti`, if a currently supported Ubuntu can't assemble the instruction, IMHO it seems like we should probably avoid it for the time being -- we otherwise don't really have stringent version requirements on the system toolchain AFAIK, just on Rust, and I've seen plenty of old Ubuntu (and other distro) installs in production in various places.

Could we `#define _bti hint #XX` and maybe the PAC instructions too for uniformity?

(I'll defer to Alex on this though if he feels differently, he's far more qualified on toolchain questions generally)

alexcrichton replied:
Oh I'm mostly just curious I don't think any concrete changes are needed. The `// actual-instr` comments are clear enough to me what it's supposed to be. If there's some assembler that needs the `hint ...` instead of the actual instruction seems fine to leave these in.

alexcrichton reviewed (approved):
Seems reasonable to me!

To confirm, have you had the chance to test this on macOS arm hardware as well? If not I'm sure @cfallin would be able to test this out.

cfallin commented:
> To confirm, have you had the chance to test this on macOS arm hardware as well? If not I'm sure @cfallin would be able to test this out.

I can confirm that `./ci/run-tests.sh` passes successfully on my M1 system.

akirilov-arm commented:
@alexcrichton I have tried to elaborate on the most important technical details in the comments - please, let me know if you feel that there is anything else that warrants further explanation. Also, now this PR fixes #4226 and, finally, I have added a fix for a minor issue that I have noticed previously - I have changed the locations of the saved FP and LR registers on the fiber stack to improve compliance with the procedure call standard.
